cloudwatch_logs_common/src/log_publisher.cpp

```diff
@@ -108,102 +107,153 @@ Aws::CloudWatchLogs::ROSCloudWatchLogsErrors LogPublisher::StopPublisherThread()
+    this->run_state_ = Aws::CloudWatchLogs::LOG_PUBLISHER_RUN_INIT_TOKEN;
+    AWS_LOG_DEBUG(__func__, "Log stream already exists, quit attempting to create it.");
+  } else {
+    AWS_LOGSTREAM_ERROR(__func__, "Failed to create log stream, status: " 
+                        << create_log_stream_status << ". Retrying ...");
+  }
+}
 
-    this->cloudwatch_facade_->GetLogStreamToken(this->log_group_, this->log_stream_, next_token);
-
-    auto shared_logs_obj = this->shared_logs_.load(std::memory_order_acquire);
-    if (nullptr != shared_logs_obj) {
-      std::list<Aws::CloudWatchLogs::Model::InputLogEvent> * logs;
-      publisher_status = shared_logs_obj->getDataAndLock(logs);
-      if (CW_LOGS_SUCCEEDED == publisher_status) {
-        if (!logs->empty()) {
-          Aws::CloudWatchLogs::ROSCloudWatchLogsErrors send_logs_status = CW_LOGS_FAILED;
-          int tries = kMaxRetries;
-
-          while (CW_LOGS_SUCCEEDED != send_logs_status && tries > 0) {
-            send_logs_status = this->cloudwatch_facade_->SendLogsToCloudWatch(
-              next_token, this->log_group_, this->log_stream_, logs);
-            if (CW_LOGS_SUCCEEDED != send_logs_status) {
-              AWS_LOG_WARN(__func__, "Unable to send logs to CloudWatch, retrying ...");
-              Aws::CloudWatchLogs::ROSCloudWatchLogsErrors get_token_status =
-                this->cloudwatch_facade_->GetLogStreamToken(this->log_group_, this->log_stream_,
-                                                            next_token);
-              if (CW_LOGS_SUCCEEDED != get_token_status) {
-                AWS_LOG_WARN(__func__,
-                             "Unable to obtain the sequence token to use, quit attempting to send "
-                             "logs to CloudWatch");
-                break;
-              }
-            }
-            tries--;
-          }
+/**
+ * Fetches the token to use for writing logs to a stream. 
+ */
+void LogPublisher::InitToken(Aws::String & next_token)
+{
+  Aws::CloudWatchLogs::ROSCloudWatchLogsErrors get_token_status =
+    this->cloudwatch_facade_->GetLogStreamToken(this->log_group_, this->log_stream_,
+                                                next_token);
+  if (CW_LOGS_SUCCEEDED == get_token_status) {
+    this->run_state_ = Aws::CloudWatchLogs::LOG_PUBLISHER_RUN_SEND_LOGS;
+  } else {
+    AWS_LOGSTREAM_ERROR(__func__, "Unable to obtain the sequence token to use, status: " 
+                        << get_token_status << ". Retrying ...");
+  }
+}
+
+/**
+ * Checks if there are logs ready to be sent out. If so then it attempts to send them
+ * to CloudWatch.
+ */
+void LogPublisher::SendLogs(Aws::String & next_token)
+{
+  Aws::CloudWatchLogs::ROSCloudWatchLogsErrors publisher_status = CW_LOGS_SUCCEEDED;
+  auto shared_logs_obj = this->shared_logs_.load(std::memory_order_acquire);
+  if (nullptr != shared_logs_obj) {
+    std::list<Aws::CloudWatchLogs::Model::InputLogEvent> * logs;
```

Comment:
It could be initialized to a `nullptr`, but it does not need to do an allocation here. This pointer is being set to reference the list in the shared object. The ownership is still considered to be the producer. 

I agree that we can probably use unique_ptrs to better show ownership, but I think it's too much scope creep for this PR.